File: algorithm/binary_search_rust/binary_search.py
from bisect import bisect_left, bisect_right

vec = [1, 5, 5, 10, 11, 11, 15]

def lower_bound(n):
    l = -1
    r = len(vec)
    while (r - l) > 1:
        mid = (l + r) // 2
        if n <= vec[mid]:
            r = mid
        else:
            l = mid
    return r

def upper_bound(n):
    l = -1
    r = len(vec)
    while (r - l) > 1:
        mid = (l + r) // 2
        if n < vec[mid]:
            r = mid
        else:
            l = mid
    return r

print(f"{vec=}")

print("\n# bisect_left")
print(f"{bisect_left(vec, -1)=}")
print(f"{bisect_left(vec, 1)=}")
print(f"{bisect_left(vec, 3)=}")
print(f"{bisect_left(vec, 5)=}")
print(f"{bisect_left(vec, 7)=}")
print(f"{bisect_left(vec, 11)=}")
print(f"{bisect_left(vec, 100)=}")

print("\n# lower_bound")
print(f"{lower_bound(-1)=}")
print(f"{lower_bound(1)=}")
print(f"{lower_bound(3)=}")
print(f"{lower_bound(5)=}")
print(f"{lower_bound(7)=}")
print(f"{lower_bound(11)=}")
print(f"{lower_bound(100)=}")

print("\n# bisect_right")
print(f"{bisect_right(vec, -1)=}")
print(f"{bisect_right(vec, 1)=}")
print(f"{bisect_right(vec, 3)=}")
print(f"{bisect_right(vec, 5)=}")
print(f"{bisect_right(vec, 7)=}")
print(f"{bisect_right(vec, 11)=}")
print(f"{bisect_right(vec, 100)=}")

print("\n# upper_bound")
print(f"{upper_bound(-1)=}")
print(f"{upper_bound(1)=}")
print(f"{upper_bound(3)=}")
print(f"{upper_bound(5)=}")
print(f"{upper_bound(7)=}")
print(f"{upper_bound(11)=}")
print(f"{upper_bound(100)=}")

File: algorithm/binary_search_rust/binary_search.rs
#![allow(non_snake_case)]

fn main() {
    let vec = vec![1, 5, 5, 10, 11, 11, 15];

    let isOK = |n: isize, m: isize| {
        n <= vec[m as usize]
    };

    // 条件を満たす最大を求める
    let lower_bound = |n: isize| {
        let mut l = -1_isize;
        let mut r = vec.len() as isize;
        while (r - l) > 1 {
            let mid = (l + r) / 2;
            if isOK(n, mid) {
                r = mid;
            } else {
                l = mid;
            }
        }
        l
    };

    // 条件を満たさない最小を求める
    let upper_bound = |n: isize| {
        let mut l = -1_isize;
        let mut r = vec.len() as isize;
        while (r - l) > 1 {
            let mid = (l + r) / 2;
            if n < vec[mid as usize] {
                r = mid;
            } else {
                l = mid;
            }
        }
        r
    };

    eprintln!("{:?}", vec);

    // lower_bound
    eprintln!("lower_bound(-1) -> {}", lower_bound(-1));
    eprintln!("lower_bound(1) -> {}", lower_bound(1));
    eprintln!("lower_bound(3) -> {}", lower_bound(3));
    eprintln!("lower_bound(5) -> {}", lower_bound(5));
    eprintln!("lower_bound(7) -> {}", lower_bound(7));
    eprintln!("lower_bound(11) -> {}", lower_bound(11));
    eprintln!("lower_bound(100) -> {}", lower_bound(100));

    // upper_bound
    eprintln!("upper_bound(-1) -> {}", upper_bound(-1));
    eprintln!("upper_bound(1) -> {}", upper_bound(1));
    eprintln!("upper_bound(3) -> {}", upper_bound(3));
    eprintln!("upper_bound(5) -> {}", upper_bound(5));
    eprintln!("upper_bound(7) -> {}", upper_bound(7));
    eprintln!("upper_bound(11) -> {}", upper_bound(11));
    eprintln!("upper_bound(100) -> {}", upper_bound(100));
}
